packages/webamp/js/components/VisPainter.ts

```diff
@@ -0,0 +1,701 @@
+export interface Vis {
+  canvas?: HTMLCanvasElement;
+  colors: string[];
+  analyser?: AnalyserNode;
+  oscStyle?: "dots" | "solid" | "lines";
+  bandwidth?: "wide" | "thin";
+  coloring?: "fire" | "line" | "normal";
+  peaks?: boolean;
+  safalloff?: "slower" | "slow" | "moderate" | "fast" | "faster";
+  sa_peak_falloff?: "slower" | "slow" | "moderate" | "fast" | "faster";
+  sa?: "analyzer" | "oscilloscope" | "none";
+}
+import { FFT } from "./FFTNullsoft";
+import {
+  renderHeight,
+  windowShade,
+  PIXEL_DENSITY,
+  doubled,
+} from "./Vis";
+
+let sapeaks = new Int16Array(76).fill(0);
+let sadata2 = new Float32Array(76).fill(0);
+let sadata = new Int16Array(76).fill(0);
+let safalloff = new Float32Array(76).fill(0);
+let sample = new Float32Array(76).fill(0);
+let barPeak = new Int16Array(76).fill(0); // Needs to be specified as Int16 else the peaks don't behave as they should
+let i: number;
+let uVar12: number;
+let falloff: number;
+let peakfalloff: number;
+
+let pushdown: number = 0;
+
+/**
+ * Base class of Visualizer (animation frame renderer engine)
+ */
+export class VisPaintHandler {
+  _vis: Vis;
+  _ctx: CanvasRenderingContext2D | null;
+
+  constructor(vis: Vis) {
+    this._vis = vis;
+    this._ctx = vis.canvas!.getContext("2d");
+  }
+
+  /**
+   * Attempt to build cached bitmaps for later use while rendering a frame.
+   * Purpose: fast rendering in animation loop
+   */
+  prepare() {}
+
+  /**
+   * Called once per frame rendering
+   */
+  paintFrame() {}
+
+  /**
+   * Attempt to cleanup cached bitmaps
+   */
+  dispose() {}
+}
+
+const SAMPLESIN = 1024;
+const SAMPLESOUT = 512;
+const fft = new FFT();
```

Comment:
Instead of constructing FFT here, try constructing it inside the constrictor of the PaintHandlers that use it and instead of assigning it to a global variable, assign it to a _property of the PaintHandler.

Reply:
Done that, though it doesn't fix the issue of the other Webamp instance being able to feed audio data to the FFT.

This is how I feed audio data to the FFT:
```ts
function processFFT(analyser: AnalyserNode, fft: FFT): void {
  const dataArray = new Uint8Array(SAMPLESIN);
  
  analyser.getByteTimeDomainData(dataArray);
  for (let i = 0; i < dataArray.length; i++) {
    inWaveData[i] = (dataArray[i] - 128) / 28;
  }
  fft.timeToFrequencyDomain(inWaveData, outSpectralData);
}```

Should I commit the change anyway?

Reply:
Changes have been committed